src/Base/Variant.h

```diff
@@ -0,0 +1,109 @@
+// *****************************************************************************
+/*!
+  \file      src/Base/Variant.h
+  \copyright 2012-2015, J. Bakosi, 2016-2017, Los Alamos National Security, LLC.
+  \brief     Variant is a wrapper around boost::variant to make it PUPable
+  \details   Variant is a wrapper around boost::variant to make it PUPable.
+
+    Since boost::variant (as well as std::variant) when default-constructed is
+    initialized to hold a value of the first alternative of its type list,
+    calling PUP that works based on a boost::visitor with a templated operator()
+    always incorrectly triggers the overload for the first type. Thus when
+    PUPing a variant not only its value but its type must also be sent during
+    migration. The class below achieves this by reading out not only the value
+    but also its zero-based index of the type alternative that is currently held
+    by the variant passed to its initializer constructor. The index and the
+    value are stored as an integer and a value in a tuple that can hold a value
+    of any of the alternative types the variant can. The index and tuple are
+    then PUPed and when unpacking, as an additional step, the variant is
+    reconstructed using the index and the value in the tuple. This latter is
+    done by invoking a runtime loop across all alternative types using
+    boost::mpl::for_each. The final variant can then be accessed by the get()
+    public member function.
+
+  \see UnitTest/tests/Base/TestPUPUtil.h or Inciter::SchemeBase.h for
+    tk::Variant in action.
+
+  \note Type-based access to tuple elements as in std::get<T>( tuple ) requires
+    C++14.
+*/
+// *****************************************************************************
+#ifndef Variant_h
+#define Variant_h
+
+#include <tuple>
+#include <boost/mpl/vector.hpp>
+
+#include "NoWarning/variant.h"
+#include "NoWarning/for_each.h"
+
+namespace tk {
+
+template< typename... Types >
+class Variant {
+
+  public:
+    //! Default constructor for Charm++ (e.g., PUP's receive side)
+    Variant() {}
+
+    //! Initializer constructor
+    //! \param[in] v boost::variant used as input/output
+    //! \details This constructor queries the index at which the variant holds a
+    //!   value and sets that value in tuple mathing the type.
+    Variant( boost::variant< Types... >& v ) : idx( v.which() ), variant(v) {
```

Comment:
Do you think it's worth abstracting (perhaps via a typedef?) to hide the implementation of the detail, to avoid explicitly stating it's a dependency on boost? Not sure if there's a compelling reason for an implementation detail like this to "leak out". When we switch to std::variant it may also make that easier

Reply:
That can be done, but I don't feel the strong need for it in this case. Though, I may not foresee enough at this point too much of a problem with this. I think at this point this is so little code, that as it is, explicitly stating boost::variant here, probably makes it more readable. Also, grepping for `boost::variant` will make switching to std::variant easy. AmI missing something?